src/js/directoryLinksProvider.js

```diff
@@ -0,0 +1,130 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+ /*globals Request, CacheTasks, async, ProviderManager, Links*/
+
+"use strict";
+
+// The const that tells where to obtain directory links
+const PREF_DIRECTORY_SOURCE = "https://tiles.services.mozilla.com/v3/links/fetch/en-US/nightly";
+
+// The frecency of a directory link
+const DIRECTORY_FRECENCY = 1000;
+
+/**
+ * Emits notifications to PlacesProvider and Links
+ */
+const DirectoryLinksProvider = {
+  _globalFetch: fetch,
+
+  _observers: new Set(),
+
+  _directoryLinksRequest: new Request(PREF_DIRECTORY_SOURCE),
+
+  /**
+   * A mapping from eTLD+1 to an enhanced link objects
+   */
+  _enhancedLinks: new Map(),
+
+  /**
+   * A mapping from site to a list of suggested link objects
+   */
+  _suggestedLinks: new Map(),
+
+  init: async(function*(mockFetch) {
+    var fetch = mockFetch || DirectoryLinksProvider._globalFetch;
+
+    Links.addObserver(DirectoryLinksProvider);
+    var response;
+    var directoryLinksInCache = yield CacheTasks.has(PREF_DIRECTORY_SOURCE, "directory_cache");
```

Comment:
Nit: "directory_cache" should probably be called "ads_cache". We are working towards being more transparent that all this stuff if related to advertisements.